url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/workflow/WorkflowEngine.java

```diff
@@ -0,0 +1,506 @@
+                    fw.write(cloudlet.getCloudletId() + indent + dft.format(cloudlet.getSubmissionTime()) + indent + dft.format(executionTime) + indent + indent + dft.format(cloudlet.getFinishTime()));
+                    
+                } catch (IOException ex) {
+                    ex.printStackTrace();
+                }
+            */
+            
+            
+        
+        
+			} else if (cloudlet.getCloudletStatus() == Cloudlet.FAILED) {
+				Log.printLine("FAILED");
+			} else if (cloudlet.getCloudletStatus() == Cloudlet.CANCELED) {
+				Log.printLine("CANCELLED");
+			}
+		}
+		long deadline = (long) (actualStartTime + execTime);
+		Log.printLine("= Deadline: "+deadline);
+		Log.printLine("= Finish time: "+endTime);
+		long makespan = (long) (endTime-actualStartTime);
+		Log.printLine("= Makespan: "+ makespan);
+		boolean violation= (endTime>deadline);
+		Log.printLine("= Violation: "+ violation);
+		Log.printLine("========== END OF SUMMARY =========");
+		Log.printLine();
+	}
+        
+        
+	public boolean hasPendingTransmissions(int vmId){
+		Iterator<WETransmission> iter = pendingTransmissions.iterator();
+		while (iter.hasNext()){
+			WETransmission tr = iter.next();
+			if (tr.getOriginId()==vmId || tr.getDestId()==vmId) return true;
+		}
+		
+		return false;
+	}
+	
+	private HashSet<Task> getTransmissionsDestinatedTo(int vmId) {
+		HashSet<Task> tasks = new HashSet<Task>();
+		
+		Iterator<WETransmission> iter = pendingTransmissions.iterator();
+		while (iter.hasNext()){
+			WETransmission tr = iter.next();
+			if (tr.getDestId()==vmId) tasks.add(tr.getTask());
+		}
+		
+		return tasks;
+	}
+	
+	private void removeOngoingTransmission(int sourceId, int destinationId, int dataId) {
+		LinkedList<WETransmission> toRemove = new LinkedList<WETransmission>();
+		
+		Iterator<WETransmission> iter = pendingTransmissions.iterator();
+		while (iter.hasNext()){
+			WETransmission tr = iter.next();
+			if (tr.getOriginId()==sourceId && tr.getDestId()==destinationId && tr.getDataId()==dataId) {
+				toRemove.add(tr);
+			}
+		}
+		
+		pendingTransmissions.removeAll(toRemove);
+	}
+
+
+     public void computeEnergy()
+     {
+         DecimalFormat dft = new DecimalFormat("#####.##");
+         TestEstimationEnergy ener = new TestEstimationEnergy();
+         String name_xml = Properties.DAG_FILE.getProperty();
+         Boolean optimized = Boolean.parseBoolean(Properties.OPTIMIZE.getProperty());
+         if(optimized)
+         {
+             String modedvfs = Properties.MODEDVFS.getProperty();
+             name_xml+="_"+modedvfs;
+         }
+         else
+             name_xml+="Fmax";
+         
+         // mode pooling ON
+         if(ener.isPooling())
+                 name_xml+="_pooling";
+         
+         double tmpEner, sumEner;
+         sumEner=0;
+            FileWriter fw = null;
+        try {
+            fw = new FileWriter("energy_" + name_xml );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fw*. The resource is referenced at lines 417, 420, 426, 445, 454. The resource is closed at line 455. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *fw* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).